### PR TITLE
[sBTC] Feat: Peg-in operation wire format

### PR DESCRIPTION
**Pull request by netrome**
This PR introduces an alpha version of the wire format for the Peg-in operation, intended for early end-to-end testing of sBTC. This version do not validate the peg wallet address, which is left to an external stacker application for the alpha test phase.

I've created #3494 as a follow-up task to add the validation of the peg wallet address once Epoch 3.0 and the updated `.pox` contract is ready.

### CODE REVIEW

**Comment by igorsyl** [comment]
Adding @donpdonp, @sergey-shandar and @xoloki as reviewers to become familiar with this code.

File: src/chainstate/burn/db/sortdb.rs
Comment (jcnelson) [inline src/chainstate/burn/db/sortdb.rs:735]: What's `address` here again?  Can you add a (sql) comment?  Thanks!

File: src/chainstate/burn/operations/mod.rs
Comment (jcnelson) [inline src/chainstate/burn/operations/mod.rs:318]: I know SIP-021's wire format appendix says that this is a Stacks address, but the text itself says that this can also be a smart contract address.  Can you make this the case?

Bytes 3-24 would be the address portion (1 byte version, 20 byte hash), and bytes 24-64 would be the name portion.

The name would be a smart contract name, so it would be between 5 and 40 bytes long.

File: testnet/stacks-node/src/tests/neon_integrations.rs
Comment (jcnelson) [inline testnet/stacks-node/src/tests/neon_integrations.rs:10426]: While what you have here is great, I just wanted to let you know that there are lighter-weight operations available for testing burnchain operations.  For example, see `test_pox_btc_ops()` and `test_stx_transfer_btc_ops()` in `src/chainstate/coordinator/tests.rs`.

File: src/chainstate/burn/db/sortdb.rs
Comment (netrome) [reply src/chainstate/burn/db/sortdb.rs:735]: Sure, it's the stacks address to receive the `sBTC`. I can add a comment.

Hmm, I think `recipient` would be a better field name throughout the code. I'll rename the field and add a SQL comment for it. Let me know if you have an even better name in mind.

File: src/chainstate/burn/operations/mod.rs
Comment (netrome) [reply src/chainstate/burn/operations/mod.rs:318]: Sure thing, I'll update.

File: testnet/stacks-node/src/tests/neon_integrations.rs
Comment (netrome) [reply testnet/stacks-node/src/tests/neon_integrations.rs:10426]: Thanks! I'll study those test cases and see if it's worth complementing the existing tests with something more lightweight to further future-proof the code. Since the neon integration tests are slow, the more code paths we can cover in lighter tests the better.

File: src/chainstate/burn/operations/mod.rs
Comment (netrome) [reply src/chainstate/burn/operations/mod.rs:318]: I'll represent the smart contract name as a String for the alpha version, leaving room for a future discussion around improved type safety for contract addresses.

The reason is that I have not found any good existing types to represent smart contract addresses. The best type I know of is `clarity::vm::types::PrincipalData`, which uses `ContractName` to validate contract names. However, this would make `chainstate::burn::operations` depend on the `clarity` crate, which I would not like to do on a whim. Also, `ContractName` can introduce unintentional panics as observed in #3431, which would risk bringing down a running node on a malformed transaction.

I guess ideally it would be best to have type alongside `StacksAddress` in `stacks_common::types::chainstate` to represent contract addresses.

Let me know any thoughts on this.

File: testnet/stacks-node/src/tests/neon_integrations.rs
Comment (netrome) [reply testnet/stacks-node/src/tests/neon_integrations.rs:10426]: I added a test case for the peg-in operation in the `chainstate::coordinator::tests` now to complement the existing test cases.

File: src/chainstate/burn/db/sortdb.rs
Comment (netrome) [reply src/chainstate/burn/db/sortdb.rs]: ✅ Done.

File: testnet/stacks-node/src/tests/neon_integrations.rs
Comment (netrome) [reply testnet/stacks-node/src/tests/neon_integrations.rs:10426]: ✅ Done

File: src/chainstate/burn/operations/mod.rs
Comment (netrome) [reply src/chainstate/burn/operations/mod.rs]: We discussed this on an sBTC working group meeting today. We're okay with depending on clarity types in `stacks_common::types::chainstate`, so I will change this ti use the `PrincipalData` type.

Comment (netrome) [reply src/chainstate/burn/operations/mod.rs]: Done :heavy_check_mark:

**Comment by netrome** [comment]
@jcnelson FYI I'm not authorized to merge this PR. Can you add me as maintainer so I can merge once it's approved and CI is green?

File: clarity/src/vm/types/mod.rs
Comment (jcnelson) [inline clarity/src/vm/types/mod.rs:1400]: Please remove this conversion. It should never be possible to convert a smart contract address into a standard address.

File: src/chainstate/burn/db/sortdb.rs
Comment (jcnelson) [inline src/chainstate/burn/db/sortdb.rs:6268]: Can you add test coverage to verify that two different burnchain blocks can contain the same transaction (i.e. with the same txid), but that `SortitionDB::get_peg_in_ops()` will return the burnchain block-specific operation metadata for a given burnchain block hash?

File: src/chainstate/burn/operations/peg_in.rs
Comment (jcnelson) [inline src/chainstate/burn/operations/peg_in.rs:92]: Please remove or use the logging facilities.

File: src/chainstate/burn/operations/peg_in.rs
Comment (jcnelson) [inline src/chainstate/burn/operations/peg_in.rs:83]: This code does not handle the case if there are trailing 0-bytes, or even the case where the `name` field is all 0's.  I think it should, because the space beyond the 64th byte can be used as a memo field for future extensions.  Instead, this code should just select out the leading non-zero bytes from the range 21 to 61, and try to decode _that_ into a ContractName.  If there are trailing 0's, it's fine -- they'll be ignored.  If the whole name field is 0, then there's no name.

Also, please don't use such long operator chains!  It's very hard to follow what this code is supposed to be doing.  Just do something like this:

```rust
let maybe_contract_name: Option<ContractName> = {
   // read to the first 0 byte
   let strlen = {
      let mut len = cmp::min(61, data.len());
      for i in 21..len {
         if data[i] == 0 {
            len = i;
            break;
          }
       }
       len
   };
      
   let name_bytes = data.get(21..strlen).unwrap_or(&[]);
   let name_string = std::str::from_utf8(&name_bytes).map_err(|_| ParseError::MalformedData)?.to_string();
   let name = ContractName::try_from(name_str).map_err(|_| ParseError::MalformedData)?;
   name
}
```

File: src/chainstate/burn/operations/peg_in.rs
Comment (jcnelson) [inline src/chainstate/burn/operations/peg_in.rs:376]: Please complete this task before merging.

File: src/chainstate/coordinator/tests.rs
Comment (jcnelson) [inline src/chainstate/coordinator/tests.rs:3411]: Please also check that it's the one you expect.

File: testnet/stacks-node/src/tests/neon_integrations.rs
Comment (jcnelson) [inline testnet/stacks-node/src/tests/neon_integrations.rs:10427]: Please also add test coverage for receiving to a contract principal.

**Comment by jcnelson** [comment]
A meta-comment.  When decoding a burnchain operation that is represented in fewer than 80 OP_RETURN bytes, you should save the remaining bytes into a `memo: Vec<u8>` and store them.  We use these "memo" bytes for in-band signaling between miners.

File: src/chainstate/burn/operations/peg_in.rs
Comment (netrome) [reply src/chainstate/burn/operations/peg_in.rs:92]: Ouch, this was unintentional. Thanks for spotting.

**Comment by netrome** [comment]
> A meta-comment. When decoding a burnchain operation that is represented in fewer than 80 OP_RETURN bytes, you should save the remaining bytes into a `memo: Vec<u8>` and store them. We use these "memo" bytes for in-band signaling between miners.

Makes sense. I'll add it here. This memo field also seems to be missing in the `PreStxOp`, `StackStxOp` and `DelegateStxOp` operations. I created #3530 as a follow-up to add it for those operations too. @jcnelson feel free to close the issue if there's any reason these ops should not have the memo field.

File: src/chainstate/burn/operations/peg_in.rs
Comment (netrome) [reply src/chainstate/burn/operations/peg_in.rs]: Alright. I factored out some pieces of the suggestion to a `leading_non_zero_bytes` utility function and updated the functionality according to your description. It should be more readable now without the long operator chain. Let me know if you have any further opinions.

File: clarity/src/vm/types/mod.rs
Comment (netrome) [reply clarity/src/vm/types/mod.rs]: Done.

File: src/chainstate/burn/operations/peg_in.rs
Comment (netrome) [reply src/chainstate/burn/operations/peg_in.rs:396]: Done. I had already added the test case, but forgot to clean up the stale todo.

File: src/chainstate/burn/db/sortdb.rs
Comment (netrome) [reply src/chainstate/burn/db/sortdb.rs]: Done. I extended the test case to submit two peg in operations with different amounts and burn header hashes.

File: src/chainstate/coordinator/tests.rs
Comment (netrome) [reply src/chainstate/coordinator/tests.rs]: Done :heavy_check_mark:

File: src/chainstate/burn/db/sortdb.rs
Comment (jcnelson) [inline src/chainstate/burn/db/sortdb.rs:747]: Hey, I know I said that this needs to be `(txid, sortition_id)` earlier, but it turns out I misspoke.  It should be `(txid, burn_header_hash)`.  It turns out that my PR #3528 is wrong as well.

Comment (jcnelson) [reply src/chainstate/burn/db/sortdb.rs:747]: I left an explanation as to why #3528 is currently mis-designed here: https://github.com/stacks-network/stacks-blockchain/pull/3528#issuecomment-1408859705

Comment (netrome) [reply src/chainstate/burn/db/sortdb.rs]: Alright, thanks for letting me know. I've updated the PR now using `(txid, burn_header_hash)` as the new primary key.

File: src/chainstate/burn/db/sortdb.rs
Comment (jcnelson) [inline src/chainstate/burn/db/sortdb.rs:732]: Please change this constant to `index_peg_in_burn_header_hash`.

File: src/chainstate/burn/operations/mod.rs
Comment (jcnelson) [inline src/chainstate/burn/operations/mod.rs:324]: This field can be more than 1 byte long. Please update the comment.

File: src/chainstate/burn/operations/peg_in.rs
Comment (jcnelson) [inline src/chainstate/burn/operations/peg_in.rs:1]: Please add the copyright statement that we use in all the other source files to this one.

File: src/chainstate/burn/operations/peg_in.rs
Comment (jcnelson) [inline src/chainstate/burn/operations/peg_in.rs:32]: N.B. This can also fail if the output is not recognized (i.e. is non-standard).

File: src/chainstate/burn/operations/peg_in.rs
Comment (jcnelson) [inline src/chainstate/burn/operations/peg_in.rs:77]: Because we know that `data.len() >= 21`, you can instead use `.expect()` here instead of `.ok_or()`.

File: src/chainstate/burn/operations/peg_in.rs
Comment (jcnelson) [inline src/chainstate/burn/operations/peg_in.rs:108]: Please add a docstring.

File: src/chainstate/burn/operations/peg_in.rs
Comment (jcnelson) [inline src/chainstate/burn/operations/peg_in.rs:109]: A defensive check that `from < to` might be appropriate here.

File: src/chainstate/burn/operations/peg_in.rs
Comment (jcnelson) [inline src/chainstate/burn/operations/peg_in.rs:112]: Bug -- should be `for i in from..end`, not 21.

File: src/chainstate/coordinator/tests.rs
Comment (jcnelson) [inline src/chainstate/coordinator/tests.rs:3242]: Can you just `.unwrap()` this so the test fails if the filesystem is inaccessible for this test?

File: testnet/stacks-node/src/tests/neon_integrations.rs
Comment (jcnelson) [inline testnet/stacks-node/src/tests/neon_integrations.rs:10516]: This is fine for now, but this information should also be obtainable from the event observer.  Please open an issue so we can remember to go back and add it.  Thanks!

**Review by jcnelson — COMMENTED** [review]
This is mostly in good shape.  Just a few minor things to do and I'll approve.  Thanks @netrome!

File: src/chainstate/burn/db/sortdb.rs
Comment (netrome) [reply src/chainstate/burn/db/sortdb.rs]: Done ✅

File: src/chainstate/burn/db/sortdb.rs
Comment (netrome) [reply src/chainstate/burn/db/sortdb.rs:2921]: Alright, I've updated `is_db_version_supported_in_epoch` now to only allow schema 5 in Epoch 2.1

File: src/chainstate/burn/operations/mod.rs
Comment (netrome) [reply src/chainstate/burn/operations/mod.rs]: Oops, thanks for spotting. Fixed!

File: src/chainstate/burn/operations/peg_in.rs
Comment (netrome) [reply src/chainstate/burn/operations/peg_in.rs:17]: Done.

File: src/chainstate/burn/operations/peg_in.rs
Comment (netrome) [reply src/chainstate/burn/operations/peg_in.rs]: Thanks for clarifying. I've updated the error message.

File: src/chainstate/burn/operations/peg_in.rs
Comment (netrome) [reply src/chainstate/burn/operations/peg_in.rs:124]: Done.

File: src/chainstate/burn/operations/peg_in.rs
Comment (netrome) [reply src/chainstate/burn/operations/peg_in.rs:130]: Makes sense. I added an assert since an error would indicate a severe bug in the code at which point a panic seems warranted. Let me know if you prefer a normal error here instead.

File: src/chainstate/burn/operations/peg_in.rs
Comment (netrome) [reply src/chainstate/burn/operations/peg_in.rs]: Thanks for spotting. Fixed.

File: src/chainstate/burn/operations/peg_in.rs
Comment (netrome) [reply src/chainstate/burn/operations/peg_in.rs]: Done.

File: src/chainstate/burn/operations/peg_in.rs
Comment (netrome) [reply src/chainstate/burn/operations/peg_in.rs:260]: Done. I've added the memo to most of the other tests as well.

File: src/chainstate/coordinator/tests.rs
Comment (netrome) [reply src/chainstate/coordinator/tests.rs:3242]: I can't simply unwrap, since that would cause the test to panic if the directory do not exist. I'm not a super fan of this pattern of ignoring results either but it's repeated in most test cases in this module so I'd recommend any changes to this pattern to be solved as a separate PR.

File: testnet/stacks-node/src/tests/neon_integrations.rs
Comment (netrome) [reply testnet/stacks-node/src/tests/neon_integrations.rs:10516]: Sure, I created #3554

**Comment by netrome** [comment]
> This is mostly in good shape. Just a few minor things to do and I'll approve. Thanks @netrome!

Thank you @jcnelson , I have addressed all your comments now. Let me know if everything looks good now so we can move this forward :rocket:

**Comment by igorsyl** [comment]
In `stx_transfer_btc_integration_test`:
```
 #5 159.5 thread 'tests::neon_integrations::stx_transfer_btc_integration_test' panicked at 'unexpected btc transaction type', testnet/stacks-node/src/tests/neon_integrations.rs:1709:21
```

**Comment by jcnelson** [comment]
cc @pavitthrap who touched this code path last

**Comment by netrome** [comment]
@jcnelson @igorsyl why was `develop` merged into this PR? This PR is based on `next`, and merging in unrelated changes makes the version history unnecessarily complex and hard to follow. I'm happy to rebase if there's any new changes in `next` that we need to get this through. Otherwise, let's try to separate refactor noise from feature additions as already agreed.

**Comment by netrome** [comment]
> In `stx_transfer_btc_integration_test`:
> 
> ```
>  #5 159.5 thread 'tests::neon_integrations::stx_transfer_btc_integration_test' panicked at 'unexpected btc transaction type', testnet/stacks-node/src/tests/neon_integrations.rs:1709:21
> ```

This is unrelated to the peg-in changes, and was introduced when `develop` was merged (which I've reverted now). This test failure needs to be addressed to unblock https://github.com/stacks-network/stacks-blockchain/pull/3564. FYI @pavitthrap

**Comment by netrome** [comment]
Rebased on latest `next` and CI is green :heavy_check_mark:. @jcnelson I'd love it if you could have a look if this is okay to merge.

**Review by jcnelson — APPROVED** [review]
LGTM!  Great job on this @netrome!

**Comment by netrome** [comment]
> LGTM! Great job on this @netrome!

Thank you! I'm not currently authorized to merge, can you press the magic button for me @jcnelson?

**Pull request merged**
